circe/src/main/scala/org/http4s/circe/middleware/JsonDebugErrorHandler.scala

```diff
@@ -0,0 +1,140 @@
+  private[this] val serviceErrorLogger =
+    org.log4s.getLogger("org.http4s.circe.jsondebugerrorhandler.service-errors")
+
+  // Can be parametric on my other PR is merged.
+  def apply[F[_]](
+      service: Kleisli[F, Request[F], Response[F]],
+      redactWhen: CaseInsensitiveString => Boolean = Headers.SensitiveHeaders.contains
+  )(implicit F: Sync[F]): Kleisli[F, Request[F], Response[F]] = Kleisli { req =>
+    import cats.implicits._
+    implicit def entEnc[M[_]: Applicative, N[_]] = JsonErrorHandlerResponse.entEnc[M, N](redactWhen)
+
+    service
+      .run(req)
+      .handleErrorWith {
+        case mf: MessageFailure =>
+          messageFailureLogger.debug(mf)(
+            s"""Message failure handling request: ${req.method} ${req.pathInfo} from ${req.remoteAddr
+              .getOrElse("<unknown>")}""")
+          mf.toHttpResponse[F](req.httpVersion).map { firstResp =>
+            Response[F](
+              status = firstResp.status,
+              httpVersion = firstResp.httpVersion,
+              headers = firstResp.headers.redactSensitive(redactWhen)
+            ).withEntity(JsonErrorHandlerResponse[F](req, mf))
+          }
+        case scala.util.control.NonFatal(t) =>
+          serviceErrorLogger.error(t)(
+            s"""Error servicing request: ${req.method} ${req.pathInfo} from ${req.remoteAddr
+              .getOrElse("<unknown>")}"""
+          )
+          Response[F](
+            Status.InternalServerError,
+            req.httpVersion,
+            Headers(
+              Connection("close".ci) ::
+                `Content-Length`.zero ::
+                Nil
+            ))
+            .withEntity(JsonErrorHandlerResponse(req, t))
+            .pure[F]
+      }
+  }
+
+  private final case class JsonErrorHandlerResponse[F[_]](
+      req: Request[F],
+      caught: Throwable
+  )
+  private object JsonErrorHandlerResponse {
+    def entEnc[F[_]: Applicative, G[_]](
+        redactWhen: CaseInsensitiveString => Boolean
+    ): EntityEncoder[F, JsonErrorHandlerResponse[G]] =
+      jsonEncoderOf(
+        Applicative[F],
+        encoder(redactWhen)
+      )
+    def encoder[F[_]](
+        redactWhen: CaseInsensitiveString => Boolean
+    ): Encoder[JsonErrorHandlerResponse[F]] = new Encoder[JsonErrorHandlerResponse[F]] {
+      def apply(a: JsonErrorHandlerResponse[F]): Json =
+        Json.obj(
+          "request" -> encodeRequest(a.req, redactWhen),
+          "throwable" -> encodeThrowable(a.caught)
+        )
+    }
+  }
+
+  private def encodeRequest[F[_]](
+      req: Request[F],
+      redactWhen: CaseInsensitiveString => Boolean): Json =
+    Json.obj(
+      "method" -> req.method.name.asJson,
+      "uri" -> Json.obj(
+        "authority" -> req.uri.authority
+          .map(
+            auth =>
+              Json.obj(
+                "host" -> auth.host.toString().asJson,
+                "port" -> auth.port.asJson,
+                "user_info" -> auth.userInfo
```

Comment:
Are we configured to drop nulls here?  I guess it doesn't really matter, but this will almost always be null.

Reply:
I can set that.